python/ray/data/tests/test_actor_pool_map_operator.py

```diff
@@ -556,6 +594,113 @@ def __call__(self, x):
         ).take_all()
 
 
+def test_actor_pool_fault_tolerance_e2e(ray_start_cluster):
+    """Test that a dataset with actor pools can finish, when
+    all nodes in the cluster are removed and added back."""
+    ray.shutdown()
+    cluster = ray_start_cluster
+    cluster.add_node(num_cpus=0)
+    ray.init()
+
+    @ray.remote(num_cpus=0)
+    class Signal:
+        def __init__(self):
+            self._node_id = ray.get_runtime_context().get_node_id()
+            self._num_alive_actors = 0
+            self._all_nodes_removed = False
+            self._all_nodes_restarted = False
+
+        async def notify_actor_alive(self):
+            self._num_alive_actors += 1
+
+        async def wait_for_actors_alive(self, value):
+            while self._num_alive_actors != value:
+                await asyncio.sleep(0.01)
+
+        async def notify_nodes_removed(self):
+            self._all_nodes_removed = True
+
+        async def notify_nodes_restarted(self):
+            self._all_nodes_restarted = True
+
+        async def wait_for_nodes_removed(self):
+            while not self._all_nodes_removed:
+                await asyncio.sleep(0.01)
+
+        async def wait_for_nodes_restarted(self):
+            while not self._all_nodes_restarted:
+                await asyncio.sleep(0.01)
+
+    # Create the signal actor on the head node.
+    signal_actor = Signal.remote()
+
+    # Spin up nodes
+    num_nodes = 1
+    nodes = []
+    for _ in range(num_nodes):
+        nodes.append(cluster.add_node(num_cpus=10))
+    cluster.wait_for_nodes()
+
+    class MyUDF:
+        def __init__(self, signal_actor):
+            self._node_id = ray.get_runtime_context().get_node_id()
+            self._signal_actor = signal_actor
+            self._signal_sent = False
+
+        def __call__(self, batch):
+            if not self._signal_sent:
+                # Notify the Actor is alive
+                self._signal_actor.notify_actor_alive.remote()
+
+                # Wait for the driver to remove nodes. This makes sure all
+                # actors are running tasks when removing nodes.
+                ray.get(self._signal_actor.wait_for_nodes_removed.remote())
+
+                # Wait for the driver to add nodes.
+                ray.get(self._signal_actor.wait_for_nodes_restarted.remote())
```

Comment:
Do we really need this latch?

Reply:
Yeah probably not needed, in that the Ray core layer would do the Actor restart transparently in this case.